tests/test_class_formatting.rs

```diff
@@ -0,0 +1,159 @@
+#![cfg(feature = "macros")]
+
+use pyo3::prelude::*;
+use pyo3::py_run;
+use std::fmt::{Display, Formatter};
+
+#[path = "../src/tests/common.rs"]
+mod common;
+
+#[pyclass(eq, str = "MyEnum.{:?}")]
+#[derive(Debug, PartialEq)]
+pub enum MyEnum {
+    Variant,
+    OtherVariant,
+}
+
+#[pyclass(eq, str)]
+#[derive(Debug, PartialEq)]
+pub enum MyEnum2 {
+    Variant,
+    OtherVariant,
+}
+
+impl Display for MyEnum2 {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+#[pyclass(eq, str = "MyEnum.{}")]
+#[derive(Debug, PartialEq)]
+pub enum MyEnum3 {
+    #[pyo3(name = "AwesomeVariant")]
+    Variant,
+    OtherVariant,
+}
+
+#[test]
+fn test_enum_class_fmt() {
+    Python::with_gil(|py| {
+        let var1 = Py::new(py, MyEnum::Variant).unwrap();
+        let var2 = Py::new(py, MyEnum2::Variant).unwrap();
+        let var3 = Py::new(py, MyEnum3::Variant).unwrap();
+        let var4 = Py::new(py, MyEnum3::OtherVariant).unwrap();
+        py_assert!(py, var1, "str(var1) == 'MyEnum.Variant'");
+        py_assert!(py, var2, "str(var2) == 'Variant'");
+        py_assert!(py, var3, "str(var3) == 'MyEnum.AwesomeVariant'");
+        py_assert!(py, var4, "str(var4) == 'MyEnum.OtherVariant'");
+    })
+}
+
+#[pyclass(str = "X: {x}, Y: {y}, Z: {z}")]
+#[derive(PartialEq, Eq, Clone, PartialOrd)]
+pub struct Point {
+    x: i32,
+    y: i32,
+    z: i32,
+}
+
+#[test]
+fn test_custom_struct_custom_str() {
+    Python::with_gil(|py| {
+        let var1 = Py::new(py, Point { x: 1, y: 2, z: 3 }).unwrap();
+        py_assert!(py, var1, "str(var1) == 'X: 1, Y: 2, Z: 3'");
+    })
+}
+
+#[pyclass(str)]
+#[derive(PartialEq, Eq, Clone, PartialOrd)]
+pub struct Point2 {
+    x: i32,
+    y: i32,
+    z: i32,
+}
+
+impl Display for Point2 {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "({}, {}, {})", self.x, self.y, self.z)
+    }
+}
+
+#[test]
+fn test_struct_str() {
+    Python::with_gil(|py| {
+        let var1 = Py::new(py, Point2 { x: 1, y: 2, z: 3 }).unwrap();
+        py_assert!(py, var1, "str(var1) == '(1, 2, 3)'");
+    })
+}
+
+#[pyclass(str = "{:?}")]
+#[derive(PartialEq, Debug)]
+enum ComplexEnumWithStr {
+    A(u32),
+    B { msg: String },
+}
```

Comment:
For complex enums we probably want individual format strings per variant, similar to how `thiserror` works on enums, or we only support the delegation to `Display` for now. From a format string like this I would have no idea what to expect for the outcome, especially if the variant have a different number of fields.

Reply:
Yes, this could be helpful.  I think it runs into the same magic-like territory of some of the other items you identified.  Ultimately, we could implement this and rely on documenting everything and hope that confused users will understand the documentation/find it, or we curtail the application of the shorthand functionality to only apply in simple cases (simple enums, structs, and no renaming used) and require explicit use of Display (and Debug when repr is eventually implemented).  I do tend to lean toward the explicit, though I certainly understand the desire for utility.